src/null-safety/migration-guide.md

```diff
@@ -1,23 +1,516 @@
+How you get the latest beta release depends on whether
+you use the Flutter SDK:
+
+* If you use the Flutter SDK,
+  switch to the beta channel:
+
+  ```terminal
+$ flutter channel beta
+$ flutter upgrade
+```
+* Otherwise, download a beta release
+  from the [Dart SDK archive][].
+
+[Dart SDK archive]: /tools/sdk/archive#beta-channel
+
+{{ site.alert.warn }}
+  If you return to working on production code, remember to
+  **switch back to a stable release**
+  (for example, by running `flutter channel stable`).
+{{ site.alert.end }}
+
+
+### Opt into null safety
+
+Setting the minimum SDK constraint to 2.12 (or a later version)
+opts the packaging into null safety by
+setting the [language version][] to at least 2.12.
+
+**Update `pubspec.yaml`** to use a 2.12 release as
+the minimum SDK constraint:
+
+```yaml
+environment:
+  sdk: '>=2.12.0-0 <3.0.0'
+```
+
+
+[language version]: /guides/language/evolution#language-versioning
+
+
+### Check dependency status
+
+Get the migration state of your package's dependencies,
+using the following command:
+
+```terminal
+$ dart pub outdated --mode=null-safety
+```
+
+If the output says **all the packages fully support null safety**,
+then you can start migrating.
+Otherwise, use the **Resolvable** column to find
+null-safe releases, if they exist.
+
+{{ site.alert.info }}
+  **Why do all dependencies need to support null safety?**
+  When all of an app's direct dependencies support null safety,
+  you can _run the app_ with sound null safety.
+  When all the dev dependencies support null safety,
+  you can _run tests_ with sound null safety.
+  You might also need null-safe dev dependencies for other reasons,
+  such as code generation.
+{{ site.alert.end }}
+
+Here's an example of the output for a simple package.
+The green checkmarked version for each package supports null safety:
+
+![Output of dart pub outdated, showing that examples_util has no releases with null safety support, but pedantic and test do](/null-safety/pub-outdated-output.png)
+
+Here are the steps to take before migrating this package to sound null safety:
+
+1. Wait for `examples_util` to be migrated.
+
+2. Update `pubspec.yaml` to use the latest compatible releases of
+   `pedantic`, `test`, and `examples_util`
+   (as listed in the **Resolvable** column).
+   For example, the new dev dependencies might look like this:
+   ```dart
+dev_dependencies:
+  examples_util: ^0.0.3-nullsafety.0
+  pedantic: ^1.10.0-nullsafety.3
+  test: ^1.16.0-nullsafety.9
+```
+
+3. Run `dart pub upgrade`.
+
+4. Run `dart pub outdated --mode=null-safety` again,
+   to make sure that all dependencies are null safe and up-to-date.
+
+
+## 2. Migrate {#step2-migrate}
+
+Most of the changes that your code needs to be null safe
+are easily predictable.
+For example, if a variable can be `null`,
+its type needs a `?` suffix.
+A named parameter that shouldn't be nullable
+needs to be marked `required`.
+
+The migration tool can make most of these changes,
```

Comment:
Perhaps add a link around the 'migration tool' text (to the heading just below), to parallel what we do for the 'migrate your code by hand' option? I think that would highlight that there are two migration paths. I'd also consider making these bullet points - that would also highlight that there are two options.